File: src/offer36.ConvertBinarySearchTree/convert_binary_search_tree.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:
    TreeNode* Convert(TreeNode* pRootOfTree)
    {
        TreeNode *pLastNode = nullptr;
        ConvertCore(pRootOfTree, &pLastNode);
        TreeNode *pHeadNode = pLastNode;
        while (pHeadNode->left != nullptr && pHeadNode != nullptr)
            pHeadNode = pHeadNode->left;
        return pHeadNode;
    }
    void ConvertCore(TreeNode* pRootOfTree, TreeNode** pLastNode)
    {
        if (pRootOfTree == nullptr)
            return;
        TreeNode* cur = pRootOfTree;
        if (cur->left != nullptr)
            ConvertCore(cur->left, pLastNode);
        
        cur->left = *pLastNode;
        if (*pLastNode != nullptr)
            (*pLastNode)->right = cur;
        
        *pLastNode = cur;
        if (cur->right != nullptr)
            ConvertCore(cur->right, pLastNode);
    }
};

File: src/offer36.ConvertBinarySearchTree/convert_binary_search_tree.java
//非递归
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
import java.util.*;
public class Solution {
    public TreeNode Convert(TreeNode pRootOfTree) {
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode cur = pRootOfTree;
        TreeNode pre = null;
        TreeNode head = null;
        while(cur != null || !stack.isEmpty()){
            while(cur != null){
                stack.push(cur);
                cur = cur.left;
            }
            if(!stack.isEmpty()){
                cur = stack.pop();
                if(pre != null){
                    pre.right = cur;
                    cur.left = pre;
                }else{
                    head = cur;
                }
                pre = cur;
                cur = cur.right;
            }
        }
       return head;
    }
}
//递归
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
import java.util.*;
public class Solution {
    TreeNode head, pre;
    public TreeNode Convert(TreeNode pRootOfTree) {
        dfs(pRootOfTree);
        return head;
    }
    private void dfs(TreeNode cur){
        if(cur == null){
            return;
        }
        dfs(cur.left);
        if(pre == null){
            head = cur;
        }else{
            pre.right = cur;
        }
        cur.left = pre;
        pre = cur;
        dfs(cur.right);
    }
}